modules/internal/ChapelBase.chpl

```diff
@@ -831,19 +831,37 @@ module ChapelBase {
     return ret;
   }
 
+  inline proc sizeof_raw(type t): size_t {
+    return __primitive("sizeof_raw", _ddata(t)):size_t;
+  }
+
   inline proc _ddata_allocate(type eltType, size: integral,
                               subloc = c_sublocid_none) {
+    pragma "insert line file info"
+      extern proc chpl_mem_array_alloc(nmemb: size_t, eltSize: size_t,
+                                       subloc: chpl_sublocID_t,
+                                       ref callPostAlloc: bool): c_void_ptr;
     var ret: _ddata(eltType);
-    var callAgain: bool;
-    __primitive("array_alloc", ret, size, subloc, c_ptrTo(callAgain), c_nil);
+    var callPostAlloc: bool;
+    const ret_voidp = chpl_mem_array_alloc(size:size_t, sizeof_raw(eltType),
+                                           subloc, callPostAlloc);
+    ret = _cast(_ddata(eltType), ret_voidp);
```

Comment:
You should be able to write this as `ret = ret_voidp : ret.type` or `ret = ret_voidp : _ddata(eltType);`